### PR TITLE
[Bugfix] [ROCm] [DSV4] [Perf] Add aiter mhc support

### PR DESCRIPTION
**Pull request by tjtanaa**
## Purpose

This PR integrates the aiter mhc kernel based on AITER `v0.1.13`, there are API changes in newer aiter versions. I will follow up with a new PR when we upgraded to new AITER version.

This PR also clean up the mhc kernels so that tilelang is lazily imported, avoid causing import error on platform that does not support tilelang.

IR Ops is not used because there seems to have some report regarding to some overhead with ir op abstraction. Kernel abstraction is also considered but since it does not manage weights nor require `preprocess_weight_after_loading` logic. So `CustomOp` is picked instead to capitalized on its hardware dispatching logic directly in the `layers/mhc.py` to avoid introducing complexity and inference overhead for now. 

This aiter mhc enablement PR is addressing the following issues

1. Main point: Optimize aiter mhc for ROCm.
    1. mhc-pre: `aiter`
    2. mhc-post `aiter`
    3. hc-head `ATOM`
2. [Critical] During integration it also bugfix the broken ROCm path caused by PR https://github.com/vllm-project/vllm/pull/41536
3. During integration it also remove the dependency to tilelang on platform that doesn't support tilelang
    - `CustomOp` is used to capitalize on its hardware platform dispatching logic, `forward_cuda` and `forward_hip`.
    - `_tilelang_ops.py` is introduced so that we can lazily import the tilelang op in `mhc/kernels/tilelang.py` function preventing throwing import error on platform that does not support tilelang.
    - Separation of ops based on its implementation approach to keep the file clean. e.g. `mhc/kernels/tilelang.py`, `mhc/kernels/torch.py`, `mhc/kernels/triton.py`, `mhc/kernels/aiter.py`

## Test Plan

Run lm-eval on DeepSeek V4 PRO on mi355x.

Command:
```
max_num_seqs=128
max_num_batched_tokens=8192
tensor_parallel_size=8
# export HF_HOME=/data/huggingface-cache
export VLLM_ROCM_USE_AITER=1
export VLLM_ROCM_USE_AITER_LINEAR=1
rm -rf /root/.cache/vllm/torch_compile_cache

MODEL=deepseek-ai/DeepSeek-V4-Pro
vllm serve ${MODEL} \
    --host localhost \
    --port 8001 \
    --dtype auto \
    --kv-cache-dtype fp8 \
    --tensor-parallel-size ${tensor_parallel_size} \
    --max-num-seqs ${max_num_seqs} \
    --distributed-executor-backend mp \
    --trust-remote-code \
    --gpu-memory-utilization 0.6 \
    --moe-backend "triton_unfused" \
    --enforce-eager \
    --tokenizer-mode "deepseek_v4" \
    --async-scheduling \
    --reasoning-parser "deepseek_v4" \
2>&1 | tee dsv40507attempt2.log
```

lmeval command

```
MODEL=deepseek-ai/DeepSeek-V4-Pro
lm_eval --model local-completions --model_args model=$MODEL,base_url=http://0.0.0.0:8001/v1/completions,num_concurrent=8,max_retries=10,max_gen_toks=2048,timeout=60000 --batch_size auto --tasks gsm8k --num_fewshot 8 --output_path . 2>&1 | tee -a eval.log
```

### Kernel Level Tests

- aiter ops are tested in aiter repo
- The hc_head_triton kernel is ported over from ATOM so a unit test is added `tests/kernels/test_mhc_kernels.py::test_hc_head_triton`

<details>
    <summary> test logs status of `hc_head_triton` (all passed) </summary>

```
WARNING 05-12 03:08:09 [config.py:70] Support for Transformers v4 is deprecated. The Transformers v4 codepath will become unmaintained in vLLM v0.22.0 and will be removed in vLLM v0.24.0. Please upgrade to Transformers v5: pip install --upgrade transformers
============================= test session starts ==============================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /app/dsv4/aitermhc
configfile: pyproject.toml
plugins: asyncio-1.3.0, anyio-4.13.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collecting ... collected 16 items

tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-4096-1] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-4096-4] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-4096-8] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-4096-128] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-7168-1] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-7168-4] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-7168-8] SKIPPED
tests/kernels/test_mhc_kernels.py::test_mhc_fused_post_pre[4-7168-128] SKIPPED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-4096-1] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-4096-4] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-4096-8] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-4096-128] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-7168-1] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-7168-4] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-7168-8] PASSED
tests/kernels/test_mhc_kernels.py::test_hc_head_triton[4-7168-128] PASSED

=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../../usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:362: 14 warnings
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 8 passed, 8 skipped, 16 warnings in 3.55s ===================
```
</details>

## Test Result

```
local-completions ({'model': 'deepseek-ai/DeepSeek-V4-Pro', 'base_url': 'http://0.0.0.0:8001/v1/completions', 'num_concurrent': 8, 'max_retries': 10, 'max_gen_toks': 2048, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 8, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|_  |0.9462|_  |0.0062|
|     |       |strict-match    |     8|exact_match|_  |0.9447|_  |0.0063|
```

## Performance Gain

| Metric | Before | After | Improvement | % Change |
|--------|--------|-------|-------------|----------|
| **Benchmark Duration** | 58.29s | 35.15s | -23.14s | **-39.7%** ⬇️ |
| **Request Throughput** | 0.34 req/s | 0.57 req/s | +0.23 req/s | **+67.6%** ⬆️ |
| **Output Token Throughput** | 34.35 tok/s | 56.96 tok/s | +22.61 tok/s | **+65.8%** ⬆️ |
| **Mean TTFT** | 1,444.51ms | 936.77ms | -507.74ms | **-35.1%** ⬇️ |
| **Mean TPOT** | 278.85ms | 167.52ms | -111.33ms | **-39.9%** ⬇️ |
| **Mean ITL** | 278.85ms | 167.52ms | -111.33ms | **-39.9%** ⬇️ |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -2309,5 +2309,186 @@ def paged_attention_common(
...
+                    hidden_size,
+                    dtype=torch.bfloat16,
+                    device=residual_flat.device,
+                ),
+            )
+
+        # AITER's Python wrapper allocates intermediate/output tensors without
+        # explicit device arguments, so run it under the residual tensor's device.
+        # else it will throw core dump error
+        with torch.device(residual_flat.device):
```
[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/_aiter_ops.py:2480]: removing that will caused OOM. So we need newer aiter version that contains this PR https://github.com/ROCm/aiter/pull/2916 which fixes this problem in AITER.

File: vllm/model_executor/models/deepseek_v4.py [hunk 2]
```diff
@@ -1340,7 +1391,7 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
             torch.empty(1, dtype=torch.float32),
             requires_grad=False,
         )
-
+        self.hc_head_op = HCHeadOp()
```
Comment (tjtanaa) [inline vllm/model_executor/models/deepseek_v4.py:1394]: it is called `hc_head_op` instead of `hc_head` because there is a name conflict with another layer property. `hc_head` already existed.

File: vllm/_aiter_ops.py [hunk 3]
```diff
@@ -2395,5 +2395,183 @@ def paged_attention_common(
...
+                ),
+                torch.empty(
+                    num_tokens,
+                    hidden_size,
+                    dtype=torch.bfloat16,
+                    device=residual_flat.device,
+                ),
+            )
+
+        # AITER's Python wrapper allocates intermediate/output tensors without
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py:2478]: This is still needed for this version of aiter. In newer aiter version, we should be able to remove this `with torch.device`. Without this, tensors will be allocated on different GPUs and causes illegal memory access.

File: vllm/_tilelang_ops.py
Comment (gnovack) [inline vllm/_tilelang_ops.py:1]: kinda curious about the choice to put the tilelang kernels into a separate `_tilelang_ops.py` file at the root. is there a functional reason for this or is it purely an organizational thing? To me it would be simpler to place these kernels in `kernels/mhc/tilelang.py` similar to what is done with triton

Comment (tjtanaa) [reply vllm/_tilelang_ops.py:1]: It is for organization and also easier to lazy load the module because tilelang kernels are like triton, we are using decorator @tilelang.jit which will fail on platform that does not support and doesn't need to install tilelang package.

I did consider the following implementation. 

`kernels/mhc/tilelang.py`
```
if current_platform.is_cuda():
    ...
    <define all of the tilelang kernels within this platform specific guard>
    ...
    
def mhc_pre(...)
   ...
   
```

But to avoid selective definition of function using platform specific guard, in the end I pick current implementation. Let me know what you think.

**Review by zyongye — APPROVED** [review]
Let's wait for the CI.

File: vllm/_tilelang_ops.py
Comment (gnovack) [reply vllm/_tilelang_ops.py:1]: Got it. yeah i am okay to move ahead with this approach
